File: haskell/Mixers.hs
{--
  Motor-mixing functions for real and simulated flight controllers
 
  Copyright (C) 2024 Simon D. Levy
 
  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, in version 3.
 
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.
 
  You should have received a copy of the GNU General Public License
  along with this program. If not, see <http://www.gnu.org/licenses/>.
--} 

{-# LANGUAGE DataKinds #-}
{-# LANGUAGE RebindableSyntax #-}

module Mixers where

import Language.Copilot
import Copilot.Compile.C99

import Demands
import Utils

runQuadX :: Demands -> (SFloat, SFloat, SFloat, SFloat)

runQuadX demands = (nw, se, sw, ne) where 

   (t, r, p, y) = getDemands demands

   nw = t - r - p  + y
   se = t - r + p  - y
   sw = t + r + p  + y
   ne = t + r - p  - y

File: attic/rust-examples/BetaFpvF405Sbus/src/mixers.rs
/*
   Copyright (c) 2023 Simon D. Levy

   This file is part of Hackflight.

   Hackflight is free software: you can redistribute it and/or modify it under
   the terms of the GNU General Public License as published by the Free
   Software Foundation, either version 3 of the License, or (at your option)
   any later version.

   Hackflight is distributed in the hope that it will be useful, but WITHOUT
   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
   FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
   more details.

   You should have received a copy of the GNU General Public License along with
   Hackflight. If not, see <https://www.gnu.org/licenses/>.
 */

use crate::datatypes::Demands;

pub fn quadxbf_mix(demands: Demands) -> (f32, f32, f32, f32) {

    let (thr, rol, pit, yaw) = 
        (demands.throttle, demands.roll, demands.pitch, demands.yaw);

    // right rear
    let m1 = thr - rol + pit - yaw; 

    // right front
    let m2 = thr - rol - pit + yaw; 

    // left rear
    let m3 = thr + rol + pit + yaw; 

    // left front
    let m4 = thr + rol - pit - yaw;

    (m1, m2, m3, m4)
}
